### PR TITLE
Incorrect state possible after retrying `ServiceDiscoverer` events

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

Clients have a configurable `serviceDiscovererRetryStrategy` to guarantee a steady stream of events to the `LoadBalancer` that never fails. It's necessary at the client level to avoid hanging requests indefinitely and let requests observe failures from ServiceDiscoverer. Also, for `PartitionedHttpClient` it's necessary to guarantee that `GroupedPublisher` never fails.

Retry is effectively a re-subscribe. According to `ServiceDiscoverer` contract (clarified in #3002), each `Subscriber` receives a "state of the world" as the first collection of events. The problem is that the state may change significantly between retries. As a result, unavailable addresses can remain inside the `LoadBalancer` forever. Example:

T1. SD delivers [a,b]
T1. LB receives [a,b]
T1. SD delivers error
T2. SD info changed ("a" got revoked)
T3. Client retries SD
T3. SD delivers [b]
T3. LB receives [b] (but still holds "a")

When we retry `ServiceDiscoverer` errors, we should keep pushing deltas downstream or purge events that are not present in the new "state of the world".

We previously had this protection but it was mistakenly removed in #1949 as part of a broader refactoring around `ServiceDiscoverer` <-> `LoadBalancer` contract.

Modifications:

- Add `RetryingServiceDiscoverer` that handles retries and keeps the state between retries.
- Use it in `DefaultSingleAddressHttpClientBuilder` and `DefaultPartitionedHttpClientBuilder`.
- Use `CastedServiceDiscoverer` to allow modifications for `ServiceDiscovererEvent` after we started to use a wildcard type in #2379.
- Pass consistent `targetResource` identifier to both `RetryingServiceDiscoverer` and `LoadBalancerFactory` to allow state correlation when inspecting heap dump.

Result:

Client keeps pushing deltas to `LoadBalancer` after retrying `ServiceDiscoverer` errors, keeping its state consistent with `ServiceDiscoverer`.

### CODE REVIEW

**Comment by idelpivnitskiy** [comment]
Marking it as "draft" because I still need to implement tests for `RetryingServiceDiscoverer`, but want to get earlier feedback for the main part before I work on tests.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    public Publisher<Collection<E>> discover(final U address) {
+        // The returned publisher is guaranteed to never fail because we retry all errors here. However, LoadBalancer
+        // can still cancel and re-subscribe in attempt to recover from unhealthy state. In this case, we need to
+        // re-initialize the ServiceDiscovererEventsCache and restart from an empty state.
+        return Publisher.defer(() -> {
+            final ServiceDiscovererEventsCache<R, E> eventsCache =
+                    new ServiceDiscovererEventsCache<>(targetResource, makeUnavailable);
+            return delegate().discover(address)
+                    .map(eventsCache::consumeAndFilter)
+                    .beforeOnError(eventsCache::errorSeen)
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:79]: Alternative approach could be to use `scanWithMapper` or `liftSync` instead of (`defer` + `map` + `beforeOnError`). I picked this one for simplicity bcz internals of `scanWithMapper` look a bit complicated for this task and `liftSync` is too low level API.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            LOGGER.debug("{} observed an error from ServiceDiscoverer", targetResource, t);
+        }
+
+        Collection<E> consumeAndFilter(final Collection<E> events) {
+            if (retainedState == NONE_RETAINED) {
+                for (E event : events) {
+                    if (UNAVAILABLE.equals(event.status())) {
+                        currentState.remove(event.address());
+                    } else {
+                        currentState.put(event.address(), event);
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:113]: Note that even though this will retain only the last event for the `address`, it should not matter because the `retainedState` will be used only to propagate `UNAVAILABLE` state that will simply remove it from LB. Any other associated meta-data (if any) doesn't matter for `UNAVAILABLE` state.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            assert currentState.isEmpty();
+            final List<E> toReturn = new ArrayList<>(events.size() + retainedState.size());
+            for (E event : events) {
+                final R address = event.address();
+                toReturn.add(event);
+                retainedState.remove(address);
+                if (!UNAVAILABLE.equals(event.status())) {
+                    currentState.put(address, event);
+                }
+            }
```
Comment (bryce-anderson) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:133]: I think we want to avoid emitting events that are already known state. In theory we know the status isn't unavailable so we just need to compare the status and metadata but we don't actually know about metadata. Since the events are coming from the same source simple equality should work to capture metadata on a refined `ServiceDiscovererEvent`.
```suggestion
            for (E event : events) {
                final R address = event.address();
                if (UNAVAILABLE.equals(event.status())) {
                    LOGGER.warn("When recovering from ServiceDiscoverer error an UNAVAILABLE event was observed " +
                            "which may indicate that the first event was not a state-of-the-world event and " +
                            "therefore is a ServiceDiscoverer bug.");
                    continue;
                }
                currentState.put(address, event);
                E oldEvent = retainedState.remove(address);
                if (!oldEvent.equals(event)) {
                    toReturn.add(event);
                }
            }
```

**Review by bryce-anderson — COMMENTED** [review]
I only got partially through this and will continue in another session.

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/RetryingServiceDiscovererTest.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class RetryingServiceDiscovererTest {
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/RetryingServiceDiscovererTest.java:61]: Recovered all tests from https://github.com/apple/servicetalk/pull/1949/files#diff-17f80381993b8c2b74a8b78674c065228c8ef16a9320fe23a11abffdfe629eb9 and enhanced those to cover more edge cases

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java (hunk 3 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:135]: The `RetryingServiceDiscoverer` should not alter the original flow of events from `ServiceDiscoverer`. We only need to purge addresses after retry to enforce new "state of the world".
I've added tests that validate duplicate events are propagated as-is and clarified that in the comment above this section.

Comment (bryce-anderson) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:135]: I think based on the contract of `ServiceDiscovererEvent` you're technically right, it is legal to do it, but it will seem _odd_ to the consumer (the LB) to receive a no-op message. That said it would mean we'd need to filter out empty results since perhaps the most common result would be the no-op result where nothing changed.

Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:135]: It won't receive a noop, it will always receive a new SOTW. See `RetryingServiceDiscovererTest.sameAddressPostRetry()`: https://github.com/apple/servicetalk/pull/3006/files#diff-43e3e0456e8b6d6ddfeeedc106c79ec7ab5a5c2cb13eb325ee91e2cd257ee4d8R104-R122

We don't know if the event changed any meta-data or not. So, after re-subscribe we have to propagate the state as-is. We only prune what is not present in the new state.

**Review by bryce-anderson — APPROVED** [review]

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 5]
```diff
@@ -839,4 +823,47 @@ private static <ResolvedAddress> HttpLoadBalancerFactory<ResolvedAddress> defaul
                 RoundRobinLoadBalancers.<ResolvedAddress, FilterableStreamingHttpLoadBalancedConnection>builder(
                                 DefaultHttpLoadBalancerFactory.class.getSimpleName()).build());
     }
+
+    // Because of the change in https://github.com/apple/servicetalk/pull/2379, we should constrain the type back to
+    // ServiceDiscovererEvent without "? extends" to allow RetryingServiceDiscoverer to mark events as UNAVAILABLE.
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:828]: It's not about bridging the type between SD and LB, it's about a possibility to pass a custom SD with custom event subtype to `HttpClients`. Without this, it won't be possible to use those implementations, they will have to change to use `ServiceDiscovererEvent` instead.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.Map;
+import java.util.function.UnaryOperator;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.client.api.ServiceDiscovererEvent.Status.UNAVAILABLE;
+import static io.servicetalk.concurrent.api.RetryStrategies.retryWithExponentialBackoffFullJitter;
+import static java.time.Duration.ofSeconds;
+import static java.util.Collections.emptyMap;
+
+final class RetryingServiceDiscoverer<U, R, E extends ServiceDiscovererEvent<R>>
```
Comment (tkountis) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:44]: should this more broadly accessible? and if-so, consider adding `shareContextOnSub` inside the defer.

**Review by tkountis — APPROVED** [review]
lgtm

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java (hunk 6 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java:44]: no, I don't have a use-case in mind where we need to share this "filter". 
for now, I will keep it without `shareContextOnSubscribe()` because subscribe can be driven from any request and I don't think the background discovery context needs to be shared with that.
